File: engine/src/main/java/de/hanno/hpengine/engine/graphics/GpuContext.java
package de.hanno.hpengine.engine.graphics;

import de.hanno.hpengine.engine.config.Config;
import de.hanno.hpengine.engine.graphics.renderer.GLU;
import de.hanno.hpengine.engine.graphics.renderer.constants.*;
import de.hanno.hpengine.engine.graphics.renderer.rendertarget.RenderTarget;
import de.hanno.hpengine.engine.graphics.state.RenderState;
import de.hanno.hpengine.engine.model.VertexBuffer;
import de.hanno.hpengine.engine.model.texture.ITexture;
import de.hanno.hpengine.engine.threads.TimeStepThread;
import de.hanno.hpengine.util.commandqueue.CommandQueue;
import de.hanno.hpengine.util.commandqueue.FutureCallable;
import org.lwjgl.opengl.GL11;

import java.nio.IntBuffer;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;
import java.util.logging.Logger;

public interface GpuContext {
    boolean CHECKERRORS = false;

    Logger LOGGER = Logger.getLogger(GpuContext.class.getName());

    RenderTarget getFrontBuffer();

    long getWindowHandle();

    int getCanvasWidth();

    int getCanvasHeight();

    void setCanvasWidth(int width);

    void setCanvasHeight(int height);

    void createNewGPUFenceForReadState(RenderState currentReadState);

    void registerPerFrameCommand(PerFrameCommandProvider perFrameCommandProvider);

    static GpuContext create() {
        Class<? extends GpuContext> gpuContextClass = Config.getInstance().getGpuContextClass();
        try {
            LOGGER.info("GpuContext is being initialized");
            GpuContext context = gpuContextClass.newInstance();
            return context;
        } catch (IllegalAccessException | InstantiationException e) {
            LOGGER.severe("GpuContext class " + gpuContextClass.getCanonicalName() + " probably doesn't feature a public no args constructor");
            e.printStackTrace();
        }
        return null;
    }

    static void exitOnGLError(String errorMessage) {
        if (!CHECKERRORS) {
            return;
        }

        int errorValue = GL11.glGetError();

        if (errorValue != GL11.GL_NO_ERROR) {
            String errorString = GLU.gluErrorString(errorValue);
            System.err.println("ERROR - " + errorMessage + ": " + errorString);

            new RuntimeException("").printStackTrace();
            System.exit(-1);
        }
    }

    boolean isError();

    void update(float seconds);

    void enable(GlCap cap);

    void disable(GlCap cap);

    void activeTexture(int textureUnitIndex);

    void bindTexture(int textureUnitIndex, GlTextureTarget target, int textureId);

    default void bindTexture(GlTextureTarget target, int textureId) {
        bindTexture(0, target, textureId);
    }

    default void bindTexture(int textureUnitIndex, ITexture<?> texture) {
        bindTexture(textureUnitIndex, texture.getTarget(), texture.getTextureId());
        texture.setUsedNow();
    }

    void bindTextures(IntBuffer textureIds);

    void bindTextures(int count, IntBuffer textureIds);

    void bindTextures(int firstUnit, int count, IntBuffer textureIds);

    default void unbindTexture(int textureUnitIndex, ITexture texture) {
        bindTexture(textureUnitIndex, texture.getTarget(), 0);
    }


    void viewPort(int x, int y, int width, int height);

    void clearColorBuffer();

    void clearDepthBuffer();

    void clearDepthAndColorBuffer();

    void bindFrameBuffer(int frameBuffer);

    void depthMask(boolean flag);

    void depthFunc(GlDepthFunc func);

    void readBuffer(int colorAttachmentIndex);

    void blendEquation(BlendMode mode);

    void blendFunc(BlendMode.Factor sfactor, BlendMode.Factor dfactor);

    void cullFace(CullMode mode);

    void clearColor(float r, float g, float b, float a);

    void bindImageTexture(int unit, int textureId, int level, boolean layered, int layer, int access, int internalFormat);

    int genTextures();

    int getAvailableVRAM();

    int getAvailableTotalVRAM();

    int getDedicatedVRAM();

    int getEvictedVRAM();

    int getEvictionCount();

    boolean isInitialized();

    void execute(Runnable runnable);

    void execute(Runnable runnable, boolean andBlock);

    <RETURN_TYPE> RETURN_TYPE calculate(Callable<RETURN_TYPE> callable);

    <RETURN_TYPE> CompletableFuture<RETURN_TYPE> execute(FutureCallable<RETURN_TYPE> command);

    long blockUntilEmpty();

    TimeStepThread getGpuThread();

    int createProgramId();

    int getMaxTextureUnits();

    CommandQueue getCommandQueue();

    void benchmark(Runnable runnable);

    void destroy();

    VertexBuffer getFullscreenBuffer();
    VertexBuffer getDebugBuffer();


    int genFrameBuffer();

    void clearCubeMap(int i, int textureFormat);

    void clearCubeMapInCubeMapArray(int textureID, int internalFormat, int width, int height, int cubeMapIndex);

    void register(RenderTarget target);

    List<RenderTarget> getRegisteredRenderTargets();
}

File: engine/src/main/java/de/hanno/hpengine/engine/graphics/GpuContext.kt
package de.hanno.hpengine.engine.graphics

import de.hanno.hpengine.engine.backend.BackendType
import de.hanno.hpengine.engine.graphics.renderer.GLU
import de.hanno.hpengine.engine.graphics.renderer.constants.*
import de.hanno.hpengine.engine.graphics.renderer.rendertarget.FrameBuffer
import de.hanno.hpengine.engine.graphics.renderer.rendertarget.RenderTarget
import de.hanno.hpengine.engine.graphics.state.RenderState
import de.hanno.hpengine.engine.vertexbuffer.VertexBuffer
import de.hanno.hpengine.engine.model.texture.Texture
import org.lwjgl.opengl.GL11
import java.nio.IntBuffer
import java.util.concurrent.Callable
import java.util.logging.Logger
import javax.vecmath.Tuple4f

interface OpenGlExecutor {
    val openGLThreadId: Long

    suspend fun <T> execute(block: () -> T): T
    operator fun <RETURN_TYPE> invoke(block: () -> RETURN_TYPE): RETURN_TYPE

    fun shutdown()
}
interface GpuContext<T: BackendType>: OpenGlExecutor {

    val backend: T

    val window: Window<T>

    val isError: Boolean

    val availableVRAM: Int

    val availableTotalVRAM: Int

    val dedicatedVRAM: Int

    val evictedVRAM: Int

    val evictionCount: Int

    val maxTextureUnits: Int

    val fullscreenBuffer: VertexBuffer
    val debugBuffer: VertexBuffer
    val sixDebugBuffers: List<VertexBuffer>

    val registeredRenderTargets: List<RenderTarget<*>>

    val features: List<GpuFeature>

    fun createNewGPUFenceForReadState(currentReadState: RenderState)

    fun update(seconds: Float)

    var cullFace: Boolean
    var cullMode: CullMode
    var depthTest: Boolean
    var blend: Boolean

    fun enable(cap: GlCap)

    fun disable(cap: GlCap)

    fun activeTexture(textureUnitIndex: Int)

    fun bindTexture(textureUnitIndex: Int, target: GlTextureTarget, textureId: Int)

    fun bindTexture(target: GlTextureTarget, textureId: Int) {
        bindTexture(0, target, textureId)
    }

    fun bindTexture(textureUnitIndex: Int, texture: Texture) {
        bindTexture(textureUnitIndex, texture.target, texture.id)
    }

    fun bindTexture(texture: Texture) {
        bindTexture(texture.target, texture.id)
    }

    fun bindTextures(textureIds: IntBuffer)

    fun bindTextures(count: Int, textureIds: IntBuffer)

    fun bindTextures(firstUnit: Int, count: Int, textureIds: IntBuffer)

    fun unbindTexture(textureUnitIndex: Int, texture: Texture) {
        bindTexture(textureUnitIndex, texture.target, 0)
    }


    fun viewPort(x: Int, y: Int, width: Int, height: Int)

    fun clearColorBuffer()

    fun clearDepthBuffer()

    fun clearDepthAndColorBuffer()

    fun bindFrameBuffer(frameBuffer: Int)

    var depthMask: Boolean

    var depthFunc: GlDepthFunc

    fun readBuffer(colorAttachmentIndex: Int)

    var blendEquation: BlendMode

    fun blendFunc(sfactor: BlendMode.Factor, dfactor: BlendMode.Factor)

    fun clearColor(r: Float, g: Float, b: Float, a: Float)
    val clearColor: Tuple4f

    fun bindImageTexture(unit: Int, textureId: Int, level: Int, layered: Boolean, layer: Int, access: Int, internalFormat: Int)

    fun genTextures(): Int

    fun createProgramId(): Int

    fun destroy()


    fun genFrameBuffer(): Int

    fun clearCubeMap(i: Int, textureFormat: Int)

    fun clearCubeMapInCubeMapArray(textureID: Int, internalFormat: Int, width: Int, height: Int, cubeMapIndex: Int)

    fun register(target: RenderTarget<*>)

    fun finishFrame(renderState: RenderState)

    fun isSupported(feature: GpuFeature): Boolean

    @JvmDefault
    fun isSupported(vararg features: GpuFeature) = isSupported(features.toList())

    @JvmDefault
    fun isSupported(features: List<GpuFeature>): SupportResult {
        features.filter { !isSupported(it) }.let { unsupportedFeatures ->
            return if(unsupportedFeatures.isEmpty()) {
                SupportResult.Supported
            } else SupportResult.UnSupported(unsupportedFeatures)
        }
    }

    fun getExceptionOnError(errorMessage: String = ""): RuntimeException?
    fun bindFrameBuffer(frameBuffer: FrameBuffer)
    fun checkCommandSyncs()

    sealed class SupportResult {
        object Supported: SupportResult()
        data class UnSupported(val unsupportedFeatures: List<GpuFeature>): SupportResult()
    }

    companion object {
        val CHECKERRORS = true

        val LOGGER = Logger.getLogger(GpuContext::class.java.name)

        @JvmStatic
        fun exitOnGLError(errorMessage: String) {
            if (!CHECKERRORS) {
                return
            }

            val errorValue = GL11.glGetError()

            if (errorValue != GL11.GL_NO_ERROR) {
                val errorString = GLU.gluErrorString(errorValue)
                System.err.println("ERROR - $errorMessage: $errorString")

                RuntimeException("").printStackTrace()
                System.exit(-1)
            }
        }
    }
}
